libp2p/builders.nim

```diff
@@ -0,0 +1,210 @@
+import
+  options, tables, chronos, bearssl,
+  switch, peerid, peerinfo, stream/connection, multiaddress,
+  crypto/crypto, transports/[transport, tcptransport],
+  muxers/[muxer, mplex/mplex],
+  protocols/[identify, secure/secure, secure/noise],
+  connmanager
+
+export
+  switch, peerid, peerinfo, connection, multiaddress, crypto
+
+type
+  SecureProtocol* {.pure.} = enum
+    Noise,
+    Secio {.deprecated.}
+
+  SwitchBuilder* {.byref.} = object
```

Comment:
I think that using a heap ref is preferable here, it's a short lived object and we can pretty much guarantee it's lifetime by setting it to nil after use.

Reply:
why would you do such thing? 
it's a small descriptor, stack is its natural place, the extra copies are just nim fault and they definitely won't change anything in this instance.
`byref` is just a bit less copying around, but please don't make me use a `ref` here.. it's just wrong.

Reply:
it's true that nim makes it clumsy but this should not make us write blasphemous code imho.

Reply:
I wish `var` return worked but it would not compile when used functionally btw, think I added a comment about it.

Reply:
Heh, I was mostly being pragmatic about avoiding potentially problematic features. If byref works I'm fine with it.

Reply:
Switched to the way you wanted it btw